# Riddler Classic 2022-04-15

https://fivethirtyeight.com/features/can-you-add-a-vowel-to-lose-a-syllable/

*This week’s Classic comes from Daniel Larsen of Bloomington, Indiana. For his research project, Daniel studied Carmichael numbers. More specifically, he proved that for a sufficiently large number N, there is guaranteed to be at least one Carmichael number between N and 2N (resembling Bertand’s postulate for prime numbers).*

*As it turns out, there’s more than one way to define Carmichael numbers. For this riddle, we’ll define N as being a Carmichael number if (1) it has no square factors, and (2) if one less than every prime factor of N is a factor of N−1.*

*That’s a lot to take in at once, so let’s look at an example. The smallest Carmichael number is 561. Sure enough, it has no square factors (other than 1, which we’re not counting here). Meanwhile, its prime factors are 3, 11 and 17. If we subtract one from each of those, we get 2, 10 and 16, all of which divide 560 (one less than the original number).*

*Daniel’s puzzle asks you to find a six-digit Carmichael number that can be written as ABCABC in base 10. (Note: The digits A, B and C are not necessarily distinct.)*

Let's start with a function that returns a set of all factors of an integer.

In [1]:
from sympy import primefactors
from functools import reduce
from math import sqrt

In [2]:
def factors(n):
        step = 2 if n % 2 else 1
        return set(reduce(list.__add__, ([i, n//i] for i in range(1, int(sqrt(n))+1, step) if n % i == 0)))

Now, let's find all the 6-digit Carmichael numbers.

In [3]:
carmichael = []

for i in range(100000, 1000000):
    is_carmichael = True
    for f in (factors(i) - set([1])):
        if (f ** 0.5).is_integer():
            is_carmichael = False
            break
    for f in primefactors(i):
        if ((i-1) % (f-1) != 0):
            is_carmichael = False
            break
    if is_carmichael:
        carmichael.append(i)

Finally, let's find all the numbers that are in the form ABCABC.

In [4]:
carmichael = [c for c in carmichael if str(c)[:3] == str(c)[3:]]

print(carmichael)

[101101]


Good, there's only one such number. That means 101101 is our answer.